File: Projects-AEB/R-AEB/cpp/src/postproc/postproc_api.cpp
// raeb_postproc.cpp

#include <memory>
#include <stdexcept>

#include "postproc_api.h"
#include "postproc_angleseg_impl.hpp"

// ------------- 全局实例（对齐 Python: g_segpost_raeb）-------------
std::unique_ptr<AngleSegPostProcessor> g_segpost_raeb = nullptr;

// ------------- 初始化 -------------
// 对齐 Python: postproc_InitRAEB(nc, nm, id2name, whitelist, conf_thresh, iou_thresh, mask_thresh, mask_up)
void postproc_InitRAEB(
    int nc,
    int nm,
    const nlohmann::json::object_t& id2name,
    const nlohmann::json::object_t& whitelist,
    float conf_thresh,
    float iou_thresh,
    float mask_thresh,
    int mask_up,
    const nlohmann::json::object_t& angle_postproc
) {
    g_segpost_raeb = std::make_unique<AngleSegPostProcessor>(
        nc,
        nm,
        id2name,
        whitelist,
        conf_thresh,
        iou_thresh,
        mask_thresh,
        mask_up,
        angle_postproc
    );
}

// ------------- 单帧后处理（只做 2D）-------------
// 对齐 Python: postproc_ProcRAEB(det_cat, proto, angle) -> objs
cv::Mat postproc_ProcRAEB(
    const cv::Mat& det_cat,
    const cv::Mat& proto,
    const cv::Mat& angle,
    bool proto_chw
) {
    if (!g_segpost_raeb) {
        throw std::runtime_error("SegPostProcessor 未初始化：请先调用 postproc_InitRAEB(...)");
    }

    // AngleSegPostProcessor::postprocess 返回：
    // objs: Nx7 CV_32F
    return g_segpost_raeb->postprocess(det_cat, proto, angle, proto_chw);
}

RaebPostprocResult postproc_ProcRAEBWithMasks(
    const cv::Mat& det_cat,
    const cv::Mat& proto,
    const cv::Mat& angle,
    bool proto_chw
) {
    if (!g_segpost_raeb) {
        throw std::runtime_error("SegPostProcessor 未初始化：请先调用 postproc_InitRAEB(...)");
    }

    auto result = g_segpost_raeb->postprocess_with_masks(det_cat, proto, angle, proto_chw);
    return RaebPostprocResult{result.objs, result.masks};
}


// ------------- 实例可视化（不带 tracking）-------------
// 对齐 Python: postproc_VisInstances(img, objs, masks, track_info=None)
cv::Mat postproc_VisInstances(
    const cv::Mat& img,
    const cv::Mat& objs,
    const cv::Mat& masks
) {
    if (!g_segpost_raeb) {
        throw std::runtime_error("SegPostProcessor 未初始化：请先调用 postproc_InitRAEB(...)");
    }
    if (img.empty() || img.type() != CV_8UC3 || img.dims != 2) {
        throw std::runtime_error("img must be (H,W,3) CV_8UC3");
    }

    return g_segpost_raeb->draw_angleins(
        img,
        objs,
        masks
    );
}

// ------------- 实例可视化（带 tracking）-------------
// 对齐 Python: postproc_VisInstances(img, objs, masks, track_info)
cv::Mat postproc_VisInstances(
    const cv::Mat& img,
    const cv::Mat& objs,
    const cv::Mat& masks,
    const cv::Mat& track_info
) {
    if (!g_segpost_raeb) {
        throw std::runtime_error("SegPostProcessor 未初始化：请先调用 postproc_InitRAEB(...)");
    }

    if (track_info.empty()) {
        return g_segpost_raeb->draw_angleins(img, objs, masks);
    }

    return g_segpost_raeb->draw_angleins_with_track_info(
        img,
        /*objs=*/objs,
        /*masks=*/masks,
        /*track_info=*/track_info
    );
}

File: Projects-AEB/R-AEB/python/visper/postproc/postproc_api.py
from __future__ import annotations

from typing import Any

import numpy as np

from .postproc_angle import AnglePostProcConfig, AnglePostProcessor
from .postproc_angleseg_impl import AngleSegPostProcessor

__all__ = ["postproc_InitRAEB", "postproc_ProcRAEB", "postproc_VisInstances"]


g_segpost_raeb: AngleSegPostProcessor | None = None
g_anglepost_raeb: AnglePostProcessor | None = None


def postproc_InitRAEB(
    nc: int,
    nm: int,
    id2name: dict[int, str],
    whitelist: dict[int, dict],
    conf_thresh: float = 0.5,
    iou_thresh: float = 0.3,
    mask_thresh: float = 0.5,
    mask_up: int = 4,
    angle_postproc_cfg: dict[str, Any] | None = None,
) -> None:
    global g_segpost_raeb, g_anglepost_raeb

    g_segpost_raeb = AngleSegPostProcessor(
        nc=nc,
        nm=nm,
        id2name=id2name,
        whitelist=whitelist,
        conf_thresh=conf_thresh,
        iou_thresh=iou_thresh,
        mask_thresh=mask_thresh,
        mask_up=mask_up,
    )
    g_anglepost_raeb = AnglePostProcessor(AnglePostProcConfig.from_dict(angle_postproc_cfg))


def postproc_ProcRAEB(
    det_cat: np.ndarray,
    proto: np.ndarray,
    angle: np.ndarray,
) -> tuple[np.ndarray, np.ndarray]:
    if g_segpost_raeb is None:
        raise RuntimeError("SegPostProcessor is not initialized. Call postproc_InitRAEB(...) first.")
    if g_anglepost_raeb is None:
        raise RuntimeError("AnglePostProcessor is not initialized. Call postproc_InitRAEB(...) first.")

    angle_decoded = g_anglepost_raeb.decode_angle(angle)
    objs, masks = g_segpost_raeb.postprocess(
        det_cat=det_cat,
        proto=proto,
        angle=angle_decoded,
    )
    objs = g_anglepost_raeb.process(objs)
    return objs, masks


def postproc_VisInstances(
    img: np.ndarray,
    objs: np.ndarray,
    masks: np.ndarray,
    track_info: np.ndarray | None = None,
) -> np.ndarray:
    if g_segpost_raeb is None:
        raise RuntimeError("SegPostProcessor is not initialized. Call postproc_InitRAEB(...) first.")

    if track_info is None:
        return g_segpost_raeb.draw_angleins(
            img=img,
            objs=objs,
            masks=masks,
        )

    return g_segpost_raeb.draw_angleins_with_track_info(
        img=img,
        objs=objs,
        masks=masks,
        track_info=track_info,
    )
